spec/app/ics-031-crosschain-queries/README.md

```diff
@@ -0,0 +1,388 @@
+
+
+    // Verify query result using the local light client of the Queried Chain. If success, then verify that the data is indeed the value associated with query.path at query.queryHeight at the Queried Chain. Otherwise, verify that query.path does not exist at query.queryHeight at the Queried Chain.
+    if (data !== null) {    
+        abortTransactionUnless(client.verifyMemership(
+            client,
+            proofHeight,
+            delayPeriodTime,
+            delayPeriodBlocks,
+            proof,
+            query.path,
+            data
+        ))
+        result = SUCCESS
+    } else {
+        abortTransactionUnless(client.verifyNonMemership(
+            client,
+            proofHeight,
+            delayPeriodTime,
+            delayPeriodBlocks,
+            proof,
+            query.path,
+        ))
+        result = FAILURE
+    }
+
+    // Delete the query from the local, private store.
+    privateStore.delete(queryPath(queryId))
+
+    // Create a query result record.
+    resultRecord = CrossChainQuery{queryIdentifier,
+                                   result,
+                                   data} 
+
+    // Store the result in the local, private store.
+    privateStore.set(resultQueryPath(queryIdentifier), resultRecord)
+
+}
+```
+- **Precondition**
+  - There exists a client with `clientId` identifier.
+  - There is a query request stored in the `privateStore` identified by `queryId`.
+- **Postcondition**
+  - The query request identified by `queryId` is deleted from the `privateStore`.
+  - The query result is stored in the `privateStore`.
+
+The `PruneCrossChainQueryResult` function is called when the caller of a query has retrieved the result and wants to delete it.
+
+```typescript
+function PruneCrossChainQueryResult(
+  queryId: Identifier,
+  queryCapability: CapabilityKey
+  ) {
+
+    // Retrieve the query result from the private store using the query's identifier.
+    resultRecord = privateStore.get(resultQueryPath(queryIdentifier))
+    abortTransactionUnless(resultRecord !== null)
+
+    // Abort the transaction unless the caller has the right to clean the query result
+    abortTransactionUnless(authenticateCapability(queryId, queryCapability))
+
+    // Delete the query result from the the local, private store.
+    privateStore.delete(resultQueryPath(queryId))
+}
+```
+- **Precondition**
+  - There is a query result stored in the `privateStore` identified by `queryId`.
+  - The caller has the right to clean the query result
+- **Postcondition**
+  - The query result identified by `queryId` is deleted from the `privateStore`.
+
+#### Timeouts
+
+Query requests have associated a `localTimeoutHeight` and a `localTimeoutTimestamp` field that specifies the height and timestamp limit at the Querying Chain after which a query is considered to have failed. 
+
+The Querying Chain calls the `checkQueryTimeout` function to check whether a specific query has timed out. 
```

Comment:
Not sure I understand how the "Querying Chain calls the `checkQueryTimeout` function". Is this a transaction? The pseudocode seems to contain several `abortTransactionUnless`.

Reply:
It depends. The spec discusses two alternatives: in one `checkQueryTimeout` is a transaction. I understand the confusion though. Ideally, this function (since it can be both a transaction or simply a function call from the module) would not include any `abortTransactionUnless`. Any suggestion on how to solve this? I can think of two solutions:
- remove `abortTransactionUnless` from `checkQueryTimeout` and use if-else conditions. Probably the best solution.
- commit to one of the discussed alternatives

Reply:
I have opted for the second solution.